lib/Sema/TypeCheckRequestFunctions.cpp

```diff
@@ -201,30 +201,44 @@ AttachedResultBuilderRequest::evaluate(Evaluator &evaluator,
 /// Attempt to infer the result builder type for a declaration.
 static Type inferResultBuilderType(ValueDecl *decl)  {
   auto dc = decl->getDeclContext();
-  if (!dc->isTypeContext() || isa<ProtocolDecl>(dc))
-    return Type();
 
   auto funcDecl = dyn_cast<FuncDecl>(decl);
-  if (!funcDecl || !funcDecl->hasBody() ||
-      !decl->getDeclContext()->getParentSourceFile())
-    return Type();
-
-  // Check whether there are any return statements in the function's body.
-  // If there are, the result builder transform will be disabled,
-  // so don't infer a result builder.
-  if (!TypeChecker::findReturnStatements(funcDecl).empty())
+  if (!funcDecl || !dc->getParentSourceFile())
     return Type();
 
-  // Only getters can have result builders. When we find one, look at
-  // the storage declaration for the purposes of witness matching.
-  auto lookupDecl = decl;
+  // For a getter, always favor the result builder type of its storage
+  // declaration if not null.
+  // For other accessors, inference is not applicable.
   if (auto accessor = dyn_cast<AccessorDecl>(funcDecl)) {
     if (accessor->getAccessorKind() != AccessorKind::Get)
       return Type();
 
-    lookupDecl = accessor->getStorage();
+    if (auto type = accessor->getStorage()->getResultBuilderType()) {
+      return type;
+    }
+  }
+
+  // Attempt inference through a dynamically replaced declaration or protocol
+  // requirements that the given declaration witnesses.
+
+  if (!dc->isTypeContext() || isa<ProtocolDecl>(dc)) {
+    return Type();
+  }
+
+  // Do not infer if the body is missing. This means that either a result
+  // builder cannot be employed, or that the parent file is a module interface
+  // and there is not way of knowing whether the omitted body suppresses
+  // inference (e.g. with a return statement).
+  if (!funcDecl->hasBody()) {
+    return Type();
   }
 
+  // Check whether there are any return statements in the function's body.
+  // If there are, the result builder transform will be disabled,
+  // so don't infer a result builder.
+  if (!TypeChecker::findReturnStatements(funcDecl).empty())
+    return Type();
```

Comment:
While here, could change this to use `BraceHasReturnRequest` since that caches the result. Probably also ought to make that request a member on AnyFunctionRef at this point.

Reply:
Thanks for the idea!

Reply:
> Probably also ought to make that request a member on AnyFunctionRef at this point.

How’s this? I took a slightly different approach.